File: kaggle_environments/envs/kore_fleets/starter_bots/java/kore/ShipyardAction.java
package kore;

public class ShipyardAction {
    public static final String SPAWN = "SPAWN";
    public static final String LAUNCH = "LAUNCH";
    public final String actionType;
    public final int shipCount;
    public final String flightPlan;

    public static ShipyardAction spawnShips(int shipCount) {
        return new ShipyardAction(SPAWN, shipCount, "");
    }

    public static ShipyardAction launchFleetWithFlightPlan(int shipCount, String flightPlan) {
        return new ShipyardAction(LAUNCH, shipCount, flightPlan);
    }

    public static ShipyardAction fromString(String raw) {
        if (raw.length() == 0) {
            throw new IllegalStateException("invalid raw shipyard empty string");
        }
        int shipCount = Integer.parseInt(raw.split("_")[1]);
        if (raw.startsWith(LAUNCH)) {
            return ShipyardAction.spawnShips(shipCount);
        }
        if (raw.startsWith(SPAWN)) {
            String flightPlan = raw.split("_")[2];
            return ShipyardAction.launchFleetWithFlightPlan(shipCount, flightPlan);
        }
        throw new IllegalStateException("invalid Shipyard Action raw " + raw);
    }

    public ShipyardAction(String type, int shipCount, String flightPlan) {
        assert type.equals(SPAWN) || type.equals(LAUNCH) : "Type must be SPAWN or LAUNCH";
        assert shipCount >= 0 : "numShips must be a non-negative number";
        this.actionType = type;
        this.shipCount = shipCount;
        this.flightPlan = flightPlan;
    }

    private boolean isSpawn() {
        return this.actionType.equals(SPAWN);
    }

    private boolean isLaunch() {
        return this.actionType.equals(LAUNCH);
    }

    public String toString() {
        if (this.isSpawn()) {
            return String.format("%s_%d", SPAWN, this.shipCount);
        }
        if (this.isLaunch()) {
            return String.format("%s_%d_%s", LAUNCH, this.shipCount, this.flightPlan);
        }
        throw new IllegalStateException("invalid Shpyard Action");
    }

}

File: kaggle_environments/envs/kore_fleets/starter_bots/ts/kore/ShipyardAction.ts

export class ShipyardAction {
    public static readonly SPAWN = "SPAWN";
    public static readonly LAUNCH = "LAUNCH";
    public readonly actionType: string;
    public readonly shipCount: number;
    public readonly flightPlan: string;

    public static spawnShips(shipCount: number): ShipyardAction {
        return new ShipyardAction(ShipyardAction.SPAWN, shipCount, "");
    }

    public static launchFleetWithFlightPlan(shipCount: number, flightPlan: string): ShipyardAction {
        return new ShipyardAction(ShipyardAction.LAUNCH, shipCount, flightPlan);
    }

    public static fromstring(raw: string): ShipyardAction {
        if (raw.length == 0) {
            throw new Error("invalid raw shipyard empty string");
        }
        const shipCount = parseInt(raw.split("_")[1]);
        if (raw.startsWith(ShipyardAction.LAUNCH)) {
            return ShipyardAction.spawnShips(shipCount);
        }
        if (raw.startsWith(ShipyardAction.SPAWN)) {
            const flightPlan = raw.split("_")[2];
            return ShipyardAction.launchFleetWithFlightPlan(shipCount, flightPlan);
        }
        throw new Error("invalid Shipyard Action raw " + raw);
    }

    public constructor(type: string, shipCount: number, flightPlan: string) {
        // assert type.equals(SPAWN) || type.equals(LAUNCH) : "Type must be SPAWN or LAUNCH";
        // assert shipCount > 0 : "numShips must be a non-negative number";
        this.actionType = type;
        this.shipCount = shipCount;
        this.flightPlan = flightPlan;
    }

    private get isSpawn(): boolean {
        return this.actionType == ShipyardAction.SPAWN;
    }

    private get isLaunch(): boolean {
        return this.actionType == ShipyardAction.LAUNCH;
    }

    public toString(): string {
        if (this.isSpawn) {
            return `${ShipyardAction.SPAWN}_${this.shipCount}`;
        }
        if (this.isLaunch) {
            return `${ShipyardAction.LAUNCH}_${this.shipCount}_${this.flightPlan}`;
        }
        throw new Error("invalid Shpyard Action");
    }

}
